pycqed/analysis_v2/timedomain_analysis.py

```diff
@@ -688,48 +987,146 @@ def unique(l):
+            - self.cal_states_dict_for_rotation: same as self.cal_states_dict
+                but ordered based on transmon state levels.
+                Will be used by self.rotate_and_project_data.
+                Ex: self.cal_states_dict =
+                        {'qb2': {'e': [-2], 'f': [-1], 'g': [-3]}}
+                    self.cal_states_dict_for_rotation =
+                        {'qb2': {'g': [-3], 'e': [-2], 'f': [-1]}}
+                If pca is to be done on a qubit, then {qbn: None}.
+                A None entry will tell rotate_and_normalize_data_IQ that pca
+                should be done in the following methods:
+                - rotate_data
+                - rotate_data_TwoD
+                - rotate_data_TwoD_same_fixed_cal_idxs
+        """
         self.cal_states_dict_for_rotation = OrderedDict()
-        states = False
         cal_states_rotations = self.cal_states_rotations
-        for key in cal_states_rotations.keys():
-            if key == 'g' or key == 'e' or key == 'f':
-                states = True
         for qbn in self.qb_names:
+            do_PCA = 'pca' in self.rotation_type[qbn].lower()
             self.cal_states_dict_for_rotation[qbn] = OrderedDict()
-            if states:
-                cal_states_rot_qb = cal_states_rotations
-            else:
-                cal_states_rot_qb = cal_states_rotations.get(qbn, [])
+            cal_states_rot_qb = cal_states_rotations.get(qbn, {})
             for i in range(len(cal_states_rot_qb)):
+                # cal state corresponding to transmon level i
                 cal_state = \
                     [k for k, idx in cal_states_rot_qb.items()
                      if idx == i][0]
-                self.cal_states_dict_for_rotation[qbn][cal_state] = \
-                    None if do_PCA and self.num_cal_points != 3 else \
-                        self.cal_states_dict[cal_state]
+                if do_PCA or not len(self.cal_states_dict[qbn]):
+                    # rotation_type is some form of pca or no cal points
+                    # information was given --> do pca
+                    self.cal_states_dict_for_rotation[qbn][cal_state] = None
+                else:
+                    # take data array index for cal_state
+                    self.cal_states_dict_for_rotation[qbn][cal_state] = \
+                        self.cal_states_dict[qbn][cal_state]
 
-    def cal_states_analysis(self):
-        self.get_cal_data_points()
+    def rotate_and_project_data(self):
+        """
+        Handles data rotation and projection based on calibration points
+        information and self.rotation_type (see class docstring for what
+        rotations types are recognized by the class).
+
+        Can handle rotation and projection based on 0, 2, 3 (not for pca)
+        calibration states.
+
+        Creates self.proc_data_dict['projected_data_dict'].
+
+        ! Note: self.data_to_fit is updated to reflect what was doe here for
```

Comment:
It was a convenience/aesthetics thing. To avoid changing data_to_fit in rotate_and_project_data I would have to have the same if/else cases as done in this method in extract_data as well and I didn't want to duplicate code. 
That being said, I see your point. Should we discuss in person how to deal with this? For now I add a comment in extract_data to point out that data_to_fit will be modified later.

Reply:
I have now cleaned this up. Do you want to have another quick look or should we merge into Proj/S17?